### ISSUE TITLE
Splitting up partisan CandidateContests

### ISSUE THREAD

**Issue by gordonje** [california-civic-data-coalition/django-calaccess-processed-data#70]
The candidate pages we scrape from CAL-ACCESS group together all candidates for the same office in the same election. This is a problem for any partisan contests (e.g., primaries before 2012) where each candidate is competing only against other candidates of the same party, rather than all the candidates at once.

Looking in the results of prior elections, published by the  Secretary of State, we can see that as late as the [June 8, 2010, primary](http://www.sos.ca.gov/elections/prior-elections/statewide-election-results/statewide-direct-primary-election-june-8-2010/statement-vote/) the percent of votes is calculated among candidates of the same party:
<img width="972" alt="2010_primary_asm_5" src="https://cloud.githubusercontent.com/assets/4304548/25957060/040af2c0-3633-11e7-8a07-3fd8e79fbe60.png">
Not so starting in the [June 5, 2012, primary](http://www.sos.ca.gov/elections/prior-elections/statewide-election-results/presidential-primary-election-june-5-2012/statement-vote/):
<img width="735" alt="2012_primary_asm_1" src="https://cloud.githubusercontent.com/assets/4304548/25956946/c11993ea-3632-11e7-9fdf-a5f372364d46.png">
We need to split these partisan contests into one for each distinct party affiliated with any candidate who ran in the primary.

**Comment by gordonje** [comment]
A little more background...

This realization came about as a result of some fixes I made to the way we group together and name elections. A unit test revealed that revealed a two-candidate discrepancy between `ScrapedCandidate` and `Candidacy`. That is, somehow in the loading it seemed like we were losing two candidates.

Turns out the problem was with State Assembly district 43, which had actually had four contests in 2010:
1. A [special primary](http://elections.cdn.sos.ca.gov/special-elections/2010-ad43/amend-results-ad43-primary-final.pdf) on April 13
2. A [special general](http://elections.cdn.sos.ca.gov/special-elections/2010-ad43/ad43-general-final.pdf) on June 8
3. A regular primary (along with [all the others](http://elections.cdn.sos.ca.gov/sov/2010-primary/pdf/101-118-ad.pdf)) also on June 8
4. A regular general (along with [all the others](http://elections.cdn.sos.ca.gov/sov/2010-general/73-state-assembly.pdf) on November 2.

This view of the results is corroborated by how the LA Times [covered the elections for this seat](http://www.latimes.com/gnp-election060910-story.html):

> When the election in November is all said and done, voters will have been asked to cast their votes four times for the same seat in about six months — a process that some stakeholders have said contributes to voter fatigue and confusion. 

Our process was conflating the two June 8 contests along with the two candidacies that were in both contests. Actually, by splitting the primaries, we should now end up with at least five contests for this office in 2010. The number of candidacies, however, should not change.

**Comment by gordonje** [comment]
Should we also split up the special primaries before the election rules changed?

I think the answer is "no".

The first special elections in which these rules applied were the [two special](http://www.sos.ca.gov/elections/prior-elections/special-elections/senate-district-28/) [state senate](http://www.sos.ca.gov/elections/prior-elections/special-elections/senate-district-17/) primaries on February 15, 2011. To these results, the SoS election division attached an [explanatory letter](http://elections.cdn.sos.ca.gov/special-elections/pdf/new-open-primary-information.pdf) explaining the rule change:
> Prior to the Top Two Candidates Open Primary Act, all candidates running in a special election, regardless of their party registration, appeared on a single Special Primary Election ballot and voters could vote for any candidate. Unless one candidate received a majority of the votes cast, the top vote-getter from each qualified party and any candidates who qualified using the independent nomination process would move on to the Special General Election.

> Now, under the Top Two Candidates Open Primary Act, all candidates running in a special election, regardless of their party preference, will still appear on a single Special Primary Election ballot and voters can vote for any candidate. However, unless one candidate receives a majority of the votes cast, the top two overall vote-getters – not the top vote-getter from each qualified party and anyone using the independent nomination process – will move on to the Special General Election.

My reading of this is that, even before the rule changed, candidates from across all parties in a special primary were in a sense competing with each other because one of them could one a majority of votes and decide who gets the seat. In these cases, there won't even be a "general" election.

Looking back at the published results for the 2010 special primary for Assembly district 43 (before the rules changed), the percent of votes cast are calculated among all candidates, not among candidates of the same party (as in pre-2012 regular primaries):
<img width="985" alt="2010_special_primary_asm_43" src="https://cloud.githubusercontent.com/assets/4304548/25958397/1bd794dc-3636-11e7-8e73-761aaacadab3.png">.

This was also true for three other pre-2011 special primaries I checked: [State Senate 37](http://elections.cdn.sos.ca.gov/special-elections/2010-sd37/final-official-results-sd37-primary.pdf), [Assembly 72](http://elections.cdn.sos.ca.gov/special-elections/2009-ad72/final-official-results-ad72-primary.pdf) and [Assembly 51](http://elections.cdn.sos.ca.gov/special-elections/2009-ad51/final-official-results-ad51-primary.pdf).

While it seems to me we probably don't want to split up the special primaries, maybe we should be linking them to the subsequent special general elections via the `runoff_for_contest_id` property. The CAL-ACCESS website use the label "SPECIAL RUNOFF" when referring to these.

tl;dr: Let's set the special elections aside and focus on splitting up the regular primaries.

**Comment by gordonje** [comment]
To pull this off, we'll need to be sure we can link each candidacy in a regular primary to a party.

Currently, we're getting the candidate's party from the `Form501Filing`. However, we can't always find a `Form501Filing` for a scraped candidate and, even when we can, the party field is sometimes blank.

Example: Jeff Perwin ran for State Senate 39 in the 2008 primary and general elections. No hyperlink for ol' Jeff on the CAL-ACCESS web pages, but we're able to dig up form501 with that name, office and primary election. That record has a filer_id (1305083), but no party.

Check out candidate page for this filer_id, though:
<img width="642" alt="perwin" src="https://cloud.githubusercontent.com/assets/4304548/25960482/ad3e3fa6-363c-11e7-9ccf-5eb90b981cf3.png">

So, somehow, CAL-ACCESS knows that this filer_id is a Republican.

**Comment by gordonje** [comment]
The `FILER_TO_FILER_TYPE_CD` table in the raw data includes a `PARTY_CD` column. There's one record in their for filer_id 1305083 with a 16002 value (`REPUBLICAN` according to `LOOKUP_CODES_CD`.

This also jives with something from a previous email from the Secretary of State's office:
> Using the FILER_TO_FILER_TYPE and F501_502 tables, the candidate and race data can be extracted.

Email also included this query:
```sql
SELECT * FROM F501_502
LEFT JOIN FILER_TO_FILER_TYPE FTFT
            ON F501_502.FILER_ID = FTFT.FILER_ID
WHERE YR_OF_ELEC = ‘2016’
AND (FTFT.FILER_ID, FTFT.EFFECT_DT) IN
            (SELECT FILER_ID, MAX(EFFECT_DT)
            FROM FILER_TO_FILER_TYPE
            GROUP BY FILER_ID)
```

The subquery here is getting the `PARTY_CD` value of the`FILER_TO_FILER_TYPE` record with the most recent `EFFECT_DT`. But we don't always want the candidate's current party, we want whatever party the candidate was affiliated with durning each election.

**Comment by gordonje** [comment]
I *think* have this additional lookup working so that I can find a party for each primary candidate. However, the unit tests aren't passing yet because our test data only includes a sample of the FILER_TO_FILER_TYPE_CD records, and these aren't covering all the scraped candidates loaded by the test.

Think the easiest way around this is to just include the entire. FILER_TO_FILER_TYPE_CD.TSV file in `example/test-data/dbwebexport.zip`.

Also, I think we can remove a bunch of the testing data cruft from the repo and gitignore it.

**Comment by gordonje** [comment]
Tests are still failing, back to having a two-candidate discrepancy between `ScrapedCandidate` and `Candidacy`.

**Comment by gordonje** [comment]
It gets weirder...

When run the commands myself (not through the tests), there's still a discrepancy, but now it is down to one candidate.

**Comment by gordonje** [comment]
So at least part of the problem is still with these four elections for Assembly District 43 in 2010. Still not seeing the special runoff created for the June 8 election. The two partisan primaries are. Looking closer again at how we coerce the scraped election names into elections...

**Comment by gordonje** [comment]
Think I figured it out. Also going to include the special election label in the contest name (since some of these contests end up under elections named "GENERAL" or "PRIMARY" because they share election dates with regular elections).

**Comment by gordonje** [comment]
So turns out, about a quarter of the pre-2012 primary candidates are missing a party:

```python
q = Candidacy.objects.filter(
    contest__election__start_time__year__lt=2012,
    contest__election__name__contains='PRIMARY'
)

float(q.filter(party__isnull=True).count()) / q.count()
```

Each of these candidacies is assigned to a split of the contest with "(UNKNOWN PARTY)"  in the name.

**Comment by palewire** [comment]
Okay. Can we make a ticket just for this problem? I could make a good code
rush type job in the future.

**Comment by gordonje** [comment]
Got it down to 20 percent with that last commit. We'll leave the rest for the other ticket.
